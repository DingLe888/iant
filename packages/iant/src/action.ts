import { TActionHandler } from './types';

export default function action(msg: string, handler: TActionHandler) {
  return () => ({
    msg,
    handler
  });
}
