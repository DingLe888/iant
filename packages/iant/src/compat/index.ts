import StoreProvider from './provider';
import withRelax from './relax';

export { StoreProvider, withRelax };
