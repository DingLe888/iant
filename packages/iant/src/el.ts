import { IQLangProps, TPath, TQLang } from './types';

let id = 0;

export class EffectLang {
  constructor(props: IQLangProps) {
    const { name, lang } = props;
    this.id = ++id;
    this.name = name;
    this.lang = lang;
  }

  private id: number;
  private name: string;
  private lang: TQLang;

  meta() {
    const lang = this.lang.slice();
    const handler = lang.pop() as Function;
    const deps = lang as Array<TPath>;

    return {
      id: this.id,
      name: this.name,
      deps,
      handler
    };
  }
}

export const EL = (lang: TQLang, name: string = '') =>
  new EffectLang({ name, lang });
