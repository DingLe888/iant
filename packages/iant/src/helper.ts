import { Store } from './store';
import { TRelaxPath } from './types';
import { isArray, isObj, isStr } from './util';

/**
 * 根据relaxPropsMapper的key做分析
 * 如果仅仅有setState或者dispatch，
 * 就不去监听store的变化
 * @param relaxPropsMapper
 */
export function isNeedRx(relaxPropsMapper: Object) {
  let keys = Object.keys(relaxPropsMapper);
  for (let key of keys) {
    if (key !== 'setState' && key !== 'dispatch') {
      return true;
    }
  }
  return false;
}

/**
 * 归集属性
 * @param relaxProps
 */
export function reduceRelaxPropsMapper(relaxProps: TRelaxPath) {
  const relaxPathMapper = {};

  for (let prop of relaxProps) {
    //如果当前的属性是数组
    //默认去最后一个字段的名称作为key
    //如果最后一位是数字，建议使用对象的方式
    if (isArray(prop)) {
      const len = prop.length;
      const last = prop[len - 1];
      relaxPathMapper[last] = prop;
    }
    //如果是字符串就直接以字符串作为key
    //如果是list.0.id这样的字符串就
    //先分割，然后走array一样的流程
    else if (isStr(prop)) {
      if (prop.indexOf('.') != -1) {
        const arr = prop.split('.');
        const len = arr.length;
        const last = arr[len - 1];
        relaxPathMapper[last] = arr;
      } else {
        relaxPathMapper[prop] = prop;
      }
    }
    //如果是对象
    else if (isObj(prop)) {
      Object.keys(prop).forEach(key => {
        let val = prop[key];
        if (isStr(val) && val.indexOf('.') != -1) {
          val = val.split('.');
        }
        relaxPathMapper[key] = val;
      });
    }
  }

  return relaxPathMapper;
}

/**
 * 计算relax路径对应的值
 * @param store
 * @param mapper
 */
export function computeRelaxProps<T>(
  store: Store,
  mapper: { [name: string]: Array<string | number> | string }
) {
  const relaxData = {
    dispatch: store.dispatch,
    setState: store.setState
  };

  for (let prop in mapper) {
    if (mapper.hasOwnProperty(prop)) {
      const val = mapper[prop];
      relaxData[prop] = store.bigQuery(val);
    }
  }

  return relaxData as T & {
    dispatch: typeof store.dispatch;
    setState: typeof store.setState;
  };
}
