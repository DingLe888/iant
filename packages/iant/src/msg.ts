import mitt from 'mitt';
import { useEffect } from 'react';

export type TMsgListeners = Array<{ name: string; handler: mitt.Handler }>;

export const msg: mitt.Emitter = new mitt();

export function useMsg(listners: TMsgListeners = []) {
  useEffect(() => {
    for (let { name, handler } of listners) {
      msg.on(name, handler);
    }
    return () => {
      for (let { name, handler } of listners) {
        msg.off(name, handler);
      }
    };
  });

  return msg;
}
