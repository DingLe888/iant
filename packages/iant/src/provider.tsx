import React from 'react';
import { StoreContext } from './context';
import { Store } from './store';
import { IProviderProps } from './types';

const noop = () => {};

export default class Provider<T = {}> extends React.Component<
  IProviderProps<T>
> {
  static defaultProps = {
    onWillMount: noop,
    onMounted: noop,
    onWillUnmount: noop
  };

  private _store: Store<T>;

  constructor(props: IProviderProps<T>) {
    super(props);
    this._store = this.props.store();

    //dev
    if (process.env.NODE_ENV !== 'production') {
      if (this._store.debug) {
        (global || window)[props.id || 'provider'] = this._store;
      }
    }
  }

  componentWillMount() {
    this.props.onWillMount(this._store);
  }

  componentDidMount() {
    this.props.onMounted(this._store);
  }

  componentDidUpdate() {
    this.props.onMounted(this._store);
  }

  render() {
    return (
      <StoreContext.Provider value={this._store}>
        {this.props.children}
      </StoreContext.Provider>
    );
  }
}
