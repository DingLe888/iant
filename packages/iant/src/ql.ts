import { IQLangProps, TPath, TQLang } from './types';

let id = 0;

export class QueryLang {
  constructor(props: IQLangProps) {
    const { name, lang } = props;
    this.id = ++id;
    this.name = name;
    this.lang = lang;
  }

  private id: number;
  private name: string;
  private lang: TQLang;

  meta() {
    const lang = this.lang.slice();
    const handler = lang.pop() as Function;
    const deps = lang as Array<TPath>;

    return {
      id: this.id,
      name: this.name,
      deps,
      handler
    };
  }
}

export const QL = (name: string, lang: TQLang) => new QueryLang({ name, lang });
