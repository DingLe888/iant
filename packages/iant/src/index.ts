import action from './action';
import { StoreProvider, withRelax } from './compat';
import { EL } from './el';
import { msg, useMsg } from './msg';
import Provider from './provider';
import { QL } from './ql';
import useRelax from './relax';
import { createStore, Store } from './store';

export {
  useRelax,
  createStore,
  Provider,
  QL,
  action,
  Store,
  useMsg,
  msg,
  EL,
  StoreProvider,
  withRelax
};
