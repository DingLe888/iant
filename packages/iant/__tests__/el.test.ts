import { MockConsole } from 'mock-jest-console';
import { createStore, EL } from '../src';

const mottEL = EL(
  [
    'count',
    'mott',
    (count: number, mott: string) => {
      console.log(mott + count);
    }
  ],
  'mottEl'
);

it('test', () => {
  const mock = new MockConsole();
  const store = createStore({
    debug: true,
    el: { mottEL },
    state: {
      count: 1,
      mott: 'build tools for human.'
    }
  })();
  store.setState(state => {
    state.count = 2;
  });
  expect(mock.logs).toMatchSnapshot();
});
