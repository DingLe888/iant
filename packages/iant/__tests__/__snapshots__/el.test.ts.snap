// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test 1`] = `
Array [
  "EL(mottEl): debug mode",
  "deps-> count",
  "val: changed(Y)",
  "deps-> mott",
  "val: changed(N)",
  "deps val was changed. trigger once",
  "build tools for human.2",
]
`;
